File: julia/problem3.jl
function greatest_prime_factor_of_(n::Int64)
    found_primes = Array{Int64}([2])
    current_prime::Int64 = 2

    remanescent_factors::Int64 = n
    greatest_prime_factor_found::Int64 = 1

    while remanescent_factors ≠ 1
        current_prime_is_a_factor_of_n = remanescent_factors % current_prime == 0
        if current_prime_is_a_factor_of_n
            greatest_prime_factor_found = current_prime
            while remanescent_factors % current_prime == 0
                remanescent_factors ÷= current_prime
            end
        end

        for i in (current_prime + 1):(n + 1)
            i_is_prime = true
            for prime in found_primes
                if i % prime == 0
                    i_is_prime = false
                    break
                end
            end
       
            if i_is_prime
                push!(found_primes, i)
                current_prime = i
                break
            end
        end
    end

    greatest_prime_factor_found
end

println(greatest_prime_factor_of_(600851475143))

File: rust/problem3.rs
fn main() {
    let n: i64 = 600851475143;
    
    let mut found_primes: Vec<i64>  = vec![2];
    let mut current_prime: i64 = 2;

    let mut remaining_factors = n;
    let mut greatest_prime_factor_found = 1;

    while remaining_factors != 1 {
        let mod_over_current_prime = remaining_factors % current_prime;
        if mod_over_current_prime == 0 {
            greatest_prime_factor_found = current_prime;
            while remaining_factors % current_prime == 0 {
                remaining_factors /= current_prime;
            }
        }

        for i in (current_prime + 1)..(n + 1) {
            let mut i_is_prime = true;
            for p in &found_primes {
                if i % p == 0 {
                    i_is_prime = false;
                    break
                }    
            }

            if i_is_prime {
                found_primes.push(i);
                current_prime = i;
                break
            }
        }
    }

    println!("Maior fator primo de {}: {}", n, greatest_prime_factor_found)
}
